File: Leetcode - C++/18_array_4Sum.cpp
// https://leetcode.com/problems/4sum/
// Time - O(N^3)

#include <vector>
#include <unordered_map>
#include <iostream>

using namespace std;

class Solution {
public:
    vector<vector<int>> fourSum(vector<int>& nums, int target) {
        vector<vector<int>> res; 
        sort(nums.begin(), nums.end());
        if(nums.size() < 4) return res; 

        for(int i = 0; i < nums.size(); i++) {
            int target3 = target - nums[i];
            for(int j = i+1; j < nums.size(); j++) {
                int target2 = target3 - nums[j];
                int front = j+1, back = nums.size()-1;
                while(front < back) {
                    int sum = nums[front] + nums[back];
                    if(sum < target2) front++;
                    else if(sum > target2) back--;
                    else {
                        vector<int> quad(4,0);
                        quad[0] = nums[i];
                        quad[1] = nums[j];
                        quad[2] = nums[front];
                        quad[3] = nums[back];
                        res.push_back(quad);

                        while(front < back && nums[front] == quad[2]) front++;
                        while(front < back && nums[back] == quad[3]) back--;
                    }
                }
                while(j+1 < nums.size() && nums[j+1] == nums[j]) j++;
            }
            while(i+1 < nums.size() && nums[i+1] == nums[i]) i++;
        }
        return res;
    }
};

int main() {
    Solution *sol = new Solution();
    vector<int> nums = {1, 0, -1, 0, -2, 2};
    int target = 0;
    vector<vector<int>> res = sol->fourSum(nums, target);
    for(int i = 0; i < res.size(); i++) {
        for(int j = 0; j < res[i].size(); j++) {
            cout << res[i][j] << " ";
        }
        cout << endl;
    }
    cout << endl;
}

File: Leetcode - Java/18_array_4Sum.java
// https://leetcode.com/problems/4sum/

class Solution {
	int len = 0;

    public List<List<Integer>> fourSum(int[] nums, int target) {
    	len = nums.length;
    	Arrays.sort(nums);
    	return ksum(nums, target, 4, 0);
    }

    private ArrayList<List<Integer>> ksum(int[] nums, int target, int k, int index) {
    	ArrayList<List<Integer>> res = new ArrayList<List<Integer>>();
    	if(index >= len) return res; 

    	if(k == 2) {
    		int i = index, j = len - 1;
    		while(i < j) {
    			if(nums[i] + nums[j] == target) {
    				List<Integer> tempList = new ArrayList<>();
    				tempList.add(nums[i]);
    				tempList.add(target-nums[i]);
    				res.add(tempList);

    				while(i < j && nums[i] == nums[i+1]) i++;
    				while(i < j && nums[j] == nums[j-1]) j--;
    				i++;
    				j--;
    			} else if(nums[i] + nums[j] < target) {
    				i++;
    			} else {
    				j--;
    			}
    		}
    	}

    	else {
    		for(int i = index; i < len - k + 1; i++) {
    			ArrayList<List<Integer>> temp = ksum(nums, target-nums[i], k-1, i+1);
    			if(temp != null) {
    				for(List<Integer> t : temp) 
    					t.add(0, nums[i]);
    				res.addAll(temp);
    			}
    			while(i < len-1 && nums[i] == nums[i+1]) i++;
    		}
    	}

    	return res;
    }
}